substrate/primitives/consensus/beefy/src/lib.rs

```diff
@@ -302,32 +307,164 @@ pub struct VoteMessage<Number, Id, Signature> {
+	/// `commitment.block_number` than commitment.payload
+	pub ancestry_proof: Option<AncestryProof<Hash>>,
+}
+
+impl<Id: RuntimeAppPublic, Header: HeaderT, Hash: HashOutput>
+	ForkEquivocationProof<Id, Header, Hash>
+{
+	fn check_fork<NodeHash: HashT<Output = Hash>>(
+		&self,
+		best_root: Hash,
+		mmr_size: u64,
+		canonical_header_hash: &Header::Hash,
+		first_mmr_block_num: Header::Number,
+		best_block_num: Header::Number,
+	) -> bool {
+		if self.commitment.block_number <= best_block_num {
+			if let Some(canonical_header) = &self.canonical_header {
+				if check_header_proof(&self.commitment, canonical_header, canonical_header_hash) {
+					// avoid verifying the ancestry proof if a valid header proof has been provided
+					return true;
+				}
+			}
+
+			if let Some(ancestry_proof) = &self.ancestry_proof {
+				return check_ancestry_proof::<Header, NodeHash>(
+					&self.commitment,
+					ancestry_proof,
+					first_mmr_block_num,
+					best_root,
+					mmr_size,
+				);
+			}
+
+			return false;
+		}
+
+		true
+	}
+
+	/// Validates [ForkEquivocationProof] with the following checks:
+	/// - if the commitment is to a block in our history, then at least a header or an ancestry
+	///   proof is provided:
+	///   - the proof is correct if the provided `canonical_header` is at height
+	/// `commitment.block_number` and `commitment.payload` != `canonical_payload(canonical_header)`
+	///   - the proof is correct if the provided `ancestry_proof` proves
+	///   `mmr_root(commitment.block_number) != mmr_root(commitment.payload)`
+	/// - `commitment` is signed by all claimed signatories
+	///
+	/// NOTE: GRANDPA finalization proof is not checked, which leads to slashing on forks. This is
+	/// fine since honest validators will not be slashed on the chain finalized by GRANDPA, which is
+	/// the only chain that ultimately matters. The only material difference not checking GRANDPA
+	/// proofs makes is that validators are not slashed for signing BEEFY commitments prior to the
+	/// blocks committed to being finalized by GRANDPA. This is fine too, since the slashing risk of
+	/// committing to an incorrect block implies validators will only sign blocks they *know* will
+	/// be finalized by GRANDPA.
+	pub fn check<MsgHash: HashT, NodeHash: HashT<Output = Hash>>(
+		&self,
+		best_root: Hash,
+		mmr_size: u64,
+		canonical_header_hash: &Header::Hash,
+		first_mmr_block_num: Header::Number,
+		best_block_num: Header::Number,
+	) -> bool
+	where
+		Id: BeefyAuthorityId<MsgHash> + PartialEq,
+	{
+		if !self.check_fork::<NodeHash>(
+			best_root,
+			mmr_size,
+			canonical_header_hash,
+			first_mmr_block_num,
+			best_block_num,
+		) {
+			return false;
+		}
+
+		return self.signatories.iter().all(|(authority_id, signature)| {
+			// TODO: refactor check_commitment_signature to take a slice of signatories
+			check_commitment_signature(&self.commitment, authority_id, signature)
+		})
```

Comment:
Moving https://github.com/paritytech/polkadot-sdk/pull/1329#discussion_r1415791979 here. Doesn't seem fixed. I think we should report each equivocated authority, even if not all signatures are valid.

Reply:
@serban300: by not fixed, you mean that while the fisherman already filters for incorrect signatures, other call sites may not?

E.g.: would you be content to just change the above to
``` rust
return self.signatories.iter().any(...)
```
?
This would then handle call sites that do not verify all their claimed equivocators actually signed the commitment.

> @Lederstrumpf: My current preference is to leave the responsibility for ensuring correct equivocation reports at the call site (i.e. the fisherman). We can also add this here, but if we are permissive in the runtime processing these, then we have to worry more about & increase our catering for DoS vectors like you point out https://github.com/paritytech/polkadot-sdk/pull/1329#discussion_r1417060121.
wdyt?

> @serban300: I don't think we can leave this check only in the fisherman. We have to check this in the pallet as well, cause otherwise we might receive equivocation reports from other callers (for example someone manually executing an extrinsic) and we would risk not validating them.

Except for manually crafted extrinsics, all other callers should be completely codified, and ensuring that all claimed signatures are valid is a simple check for these.
For manually crafted extrinsics, I fully agree with you that there is a meaningful risk that this extrinsic is not crafted correctly. In this case however, if the caller sees their report rejected, they can simply resubmit (if we're already at the stage of manually crafting extrinsics, both the additional cost & time requirements are negligible).

---

Overall, I recommend having a principled design here:
1. Put the burden of ensuring correct reports on the reporter to minimize DoS vectors.
This is the preferred route if call sites craft only correct reports, but induces latency on miscrafted reports (an active observer would have to pick up the report getting rejected and craft a correct extrinsic to replace it, or fix & retrigger the call site).
2. Put the burden of ensuring correct reports on the runtime and accept potential DoS vectors since callers can mix gunk into valid reports that the runtime must now filter (and potentially reward too).
This is the preferred route if we want to cater for faulty call sites. Note a caveat of this route that you're shifting complexity downstream since you require more checks to ensure you're not accidentally slashing honest validators - that's avoided by only allowing equivocation reports fully correct wrt. claimed signatories.

Reply:
I'm not sure what's the right thing to do on the runtime side. What reports we should accept or reject there. Still thinking. But at least for the reporting code in the beefy client:

```
if crate::justification::verify_with_validator_set::<B>(
  				commitment.block_number,
  				&validator_set,
  				&BeefyVersionedFinalityProof::<B>::V1(signed_commitment),
  			)
```
We only report if 2/3 of the signatures are correct. Shouldn't we report even if 1 signature is correct ? I might be wrong. I remember I have been confused about this part before.